pkg/virt-launcher/virtwrap/cli/event.go

```diff
@@ -0,0 +1,85 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2021 Red Hat, Inc.
+ *
+ */
+
+package cli
+
+import (
+	"fmt"
+
+	"libvirt.org/libvirt-go"
+)
+
+type DomainEventDeviceRemovedRegistrar struct {
+	connect        *libvirt.Connect
+	domain         *libvirt.Domain
+	registrationID int
+}
+
+func NewDomainEventDeviceRemovedRegistrar(connection Connection, virDomain VirDomain) *DomainEventDeviceRemovedRegistrar {
+	libvirtConnect := extractLibvirtConnect(connection)
+	libvirtDomain := extractLibvirtDomain(virDomain)
+	if libvirtConnect != nil && libvirtDomain != nil {
+		return &DomainEventDeviceRemovedRegistrar{
+			connect: libvirtConnect,
+			domain:  libvirtDomain,
+		}
+	}
+	return nil
+}
+
```

Comment:
I think we can generalize those a bit more so it will make sense using it for other events
What you think about renaming to LibvirtDomainEvenRegistrar?

Reply:
Current implementation is specific to "DeviceRemoved" as per the methods below.
In order to generalize it, there is a need to crate a new struct and embed it into this one.
e.g.:
```
type DomainEventDeviceRemovedRegistrar struct {
    LibvirtDomainEvenRegistrar
    registrationID int
}
```

Therefore, if the need will present itself to re-use this for other events, adding this should not be a problem and no changes to the public API is needed. (creating a new type and a factory that the current ones can use)

Reply:
You already created a new package for handling libvirt events, we would probably want to reuse it for other events as well
I think the implementation should be generic enough in the first place

About warping it  with DomainEventDeviceRemovedRegistrar, I think renaming be enough
because there is no concrete logic for the detach command the only thing that makes it concrete is the naming
Did I missed something?

Reply:
I cannot predict the future and did not saw it used for something else at this time.
The only best thing I can do is to make sure it is correct now and that it does not need to change if a generic solution will be needed in the future.

> About warping it with DomainEventDeviceRemovedRegistrar, I think renaming be enough
> because there is no concrete logic for the detach command the only thing that makes it concrete is the naming
> Did I missed something?

You missed that the structure and methods bind to it are specific to the DeviceRemove event and not to anything generic.
We create a registrar object of a concrete DeviceRemove, it has Register and Deregister methods for this specific event.
If you have a better solution, please give it in full here so we can see if it fits. As it stands now, renaming the type and factory to a generic name will be incorrect.